File: hostagent/internal/api/runs.go
package api

import (
	"net/http"
	"strconv"

	"orchestra/hostagent/internal/store"
)

// handleRunRecord persists a manually-launched agent run (from Delivery) so it
// joins the run history and the optimization loop.
func (s *Server) handleRunRecord(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Name        string `json:"name"`
		Repo        string `json:"repo"`
		Branch      string `json:"branch"`
		Task        string `json:"task"`
		Template    string `json:"template"`
		TemplateRef string `json:"templateRef"`
		ContainerID string `json:"containerId"`
		RunID       string `json:"runId"`
	}
	if !decode(w, r, &req) {
		return
	}
	id, err := s.store.RecordAgentRun(store.AgentRun{
		Source: "manual", Name: req.Name, Repo: req.Repo, Branch: req.Branch, Task: req.Task,
		Template: req.Template, TemplateRef: req.TemplateRef, ContainerID: req.ContainerID, RunID: req.RunID,
	})
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 201, map[string]int64{"id": id})
}

// handleRunList returns recent manual agent runs.
func (s *Server) handleRunList(w http.ResponseWriter, r *http.Request) {
	limit := 0
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			limit = n
		}
	}
	runs, err := s.store.AgentRuns(limit)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if runs == nil {
		runs = []store.AgentRun{}
	}
	writeJSON(w, 200, map[string]any{"runs": runs})
}

File: src/lib/runs.ts
// Client for the host agent's manual agent-run history: record a run launched
// from Delivery and list the recent ones, so manual runs share the run-history
// + optimization loop with scheduled runs.

const BASE = import.meta.env.VITE_HOSTAGENT_URL ?? "http://127.0.0.1:8788";

export interface AgentRun {
  id: number;
  source: string;
  name: string;
  repo: string;
  branch: string;
  task: string;
  template: string;
  templateRef: string;
  containerId: string;
  runId: string;
  createdAt: string;
}

export interface RunRecordInput {
  name: string;
  repo: string;
  branch: string;
  task?: string;
  template?: string;
  templateRef?: string;
  containerId?: string;
  runId?: string;
}

async function req<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(BASE + path, {
    ...init,
    headers: { "Content-Type": "application/json", ...(init?.headers ?? {}) },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error((body as { error?: string }).error ?? `HTTP ${res.status}`);
  }
  return res.json() as Promise<T>;
}

export const runs = {
  /** Persist a manually-launched run. Best-effort; callers ignore failures. */
  record: (r: RunRecordInput) => req<{ id: number }>("/runs", { method: "POST", body: JSON.stringify(r) }),
  /** Recent manual runs, newest first. */
  list: (limit?: number) => req<{ runs: AgentRun[] }>(`/runs${limit ? `?limit=${limit}` : ""}`).then((r) => r.runs),
};
